File: java/lib/SnobotSim/MotorSim/SimpleMotorSimulator.cpp
/*
 * SimpleMotorSimulator.cpp
 *
 *  Created on: May 5, 2017
 *      Author: PJ
 */

#include "SnobotSim/MotorSim/SimpleMotorSimulator.h"

SimpleMotorSimulator::SimpleMotorSimulator(double aMaxSpeed) :
        mMaxSpeed(aMaxSpeed),
        mVoltagePercent(0),
        mVelocity(0),
        mPosition(0)
{

}

SimpleMotorSimulator::~SimpleMotorSimulator()
{

}

void SimpleMotorSimulator::setVoltagePercentage(double aSpeed)
{
    mVoltagePercent = aSpeed;
}

double SimpleMotorSimulator::getVoltagePercentage()
{
    return mVoltagePercent;
}

double SimpleMotorSimulator::getAcceleration()
{
    return 0;
}

double SimpleMotorSimulator::getVelocity()
{
    return mVelocity;
}

double SimpleMotorSimulator::getPosition()
{
    return mPosition;
}

void SimpleMotorSimulator::reset()
{
    reset(0, 0, 0);
}

void SimpleMotorSimulator::reset(double aPosition, double aVelocity, double aCurrent)
{
    mPosition = aPosition;
    mVelocity = aVelocity;
}

void SimpleMotorSimulator::update(double aUpdateTime)
{
    mVelocity = mMaxSpeed * mVoltagePercent;
    mPosition += mVelocity * aUpdateTime;
}

File: snobot_sim_java/src/main/java/com/snobot/simulator/motor_sim/SimpleMotorSimulator.java
package com.snobot.simulator.motor_sim;

import com.snobot.simulator.module_wrapper.interfaces.IMotorSimulator;

public class SimpleMotorSimulator implements IMotorSimulator
{
    private final SimpleMotorSimulationConfig mConfig;

    private double mVoltagePercent;
    private double mVelocity;
    private double mPosition;

    public SimpleMotorSimulator(SimpleMotorSimulationConfig aConfig)
    {
        mConfig = aConfig;
    }

    @Override
    public void setVoltagePercentage(double aSpeed)
    {
        mVoltagePercent = aSpeed;
    }

    @Override
    public double getVoltagePercentage()
    {
        return mVoltagePercent;
    }

    @Override
    public double getAcceleration()
    {
        return 0;
    }

    @Override
    public double getVelocity()
    {
        return mVelocity;
    }

    @Override
    public double getPosition()
    {
        return mPosition;
    }

    @Override
    public double getCurrent()
    {
        return 0;
    }

    @Override
    public void reset()
    {
        reset(0, 0, 0);
    }

    @Override
    public void reset(double aPosition, double aVelocity, double aCurrent)
    {
        mPosition = aPosition;
        mVelocity = aVelocity;
    }

    @Override
    public void update(double aUpdateTime)
    {
        mVelocity = mConfig.mMaxSpeed * mVoltagePercent;
        mPosition += mVelocity * aUpdateTime;
    }

    public SimpleMotorSimulationConfig getConfig()
    {
        return mConfig;
    }
}
